src/libpmemset/sds.c

```diff
@@ -0,0 +1,351 @@
+ */
+struct pmemset_sds_record *
+pmemset_sds_find_record(struct pmem2_map *map)
+{
+	struct pmemset_sds_record sds_record;
+	sds_record.map = map;
+
+	struct ravl_node *node;
+
+	util_rwlock_rdlock(&State.sds_lock);
+	node = ravl_find(State.rtree, &sds_record, RAVL_PREDICATE_EQUAL);
+	util_rwlock_unlock(&State.sds_lock);
+
+	if (!node)
+		return NULL;
+
+	return ravl_data(node);
+}
+
+/*
+ * pmemset_sds_record_get_sds -- get SDS from sds record
+ */
+struct pmemset_sds *
+pmemset_sds_record_get_sds(struct pmemset_sds_record *sds_record)
+{
+	return sds_record->sds;
+}
+
+/*
+ * pmemset_sds_record_get_source -- get source from sds record
+ */
+struct pmemset_source *
+pmemset_sds_record_get_source(struct pmemset_sds_record *sds_record)
+{
+	return sds_record->src;
+}
+
+/*
+ * pmemset_sds_register_record -- register sds record in the sds tree
+ */
+int
+pmemset_sds_register_record(struct pmemset_sds *sds, struct pmemset_source *src,
+		struct pmem2_map *p2map)
+{
+	struct pmemset_sds_record sds_record;
+	sds_record.sds = sds;
+	sds_record.src = src;
+	sds_record.map = p2map;
+
+	util_rwlock_wrlock(&State.sds_lock);
+	sds->refcount += 1;
+	int ret = ravl_emplace_copy(State.rtree, &sds_record);
+	if (ret)
+		goto err_lower_ref;
+
+	util_rwlock_unlock(&State.sds_lock);
+	return 0;
+
+err_lower_ref:
+	sds->refcount -= 1;
+	util_rwlock_unlock(&State.sds_lock);
+	return ret;
+}
+
+/*
+ * pmemset_sds_unregister_record -- unregister sds record from the sds tree
+ */
+int
+pmemset_sds_unregister_record(struct pmemset_sds_record *sds_record)
+{
+	int ret = 0;
+	struct pmemset_sds *sds = sds_record->sds;
+
+	util_rwlock_wrlock(&State.sds_lock);
+	struct ravl_node *node = ravl_find(State.rtree, sds_record,
+			RAVL_PREDICATE_EQUAL);
+	if (!node) {
+		ret = -ENOENT;
+		goto err_unlock;
+	}
+
+	ravl_remove(State.rtree, node);
+
+	sds->refcount -= 1;
+	ASSERT(sds->refcount >= 0);
+
+err_unlock:
+	util_rwlock_unlock(&State.sds_lock);
+	return ret;
+}
+
+/*
+ * pmemset_sds_new -- allocates and initialize SDS structure
+ */
+int
+pmemset_sds_new(struct pmemset_sds **sds_ptr, struct pmemset_source *src)
+{
+	LOG(3, "sds_ptr %p src %p", sds_ptr, src);
+
+	*sds_ptr = NULL;
+	struct pmemset_file *part_file = pmemset_source_get_set_file(src);
+	struct pmem2_source *pmem2_src =
+			pmemset_file_get_pmem2_source(part_file);
+
+	int ret = 0;
+	struct pmemset_sds *sds = pmemset_malloc(sizeof(*sds), &ret);
+	if (ret)
+		return ret;
+
+	memset(sds->id, 0, PMEMSET_SDS_DEVICE_ID_LEN);
+
+	/* read device unsafe shutdown count */
+	ret = pmem2_source_device_usc(pmem2_src, &sds->usc);
+	if (ret) {
+		if (ret == PMEM2_E_NOSUPP)
+			ret = PMEMSET_E_SDS_ENOSUPP;
+		goto err_free_sds;
+	}
+
+	/* read device ID length */
+	size_t len = 0;
+	ret = pmem2_source_device_id(pmem2_src, NULL, &len);
+	if (ret) {
+		if (ret == PMEM2_E_NOSUPP)
+			ret =  PMEMSET_E_SDS_ENOSUPP;
+		goto err_free_sds;
+	}
+
+	ASSERT(len <= PMEMSET_SDS_DEVICE_ID_LEN);
```

Comment:
Nah, this can technically speaking happen on some bizarre configurations. Handle this error normally.